File: Cpp-SDL2/bubble_text.cpp
#include "bubble_text.h"
#include "game.h"

BubbleText::BubbleText(SDL_Renderer* renderer, const char* text, int text_size, float radius, SDL_Color outer_color, SDL_Color center_color)
    : renderer(renderer) {

    // create font based on a true type font and size.
    TTF_Font *font = TTF_OpenFont("fonts/freesansbold.ttf", text_size);
    if (!font) {
        auto error = fmt::format("Error creating font: {}", TTF_GetError());
        throw std::runtime_error(error);
    }

    // Create initial Text to calculate target surface size.
    SDL_Surface *text_surf = TTF_RenderText_Blended(font, text, outer_color);
    if (!text_surf) {
        auto error = fmt::format("Error creating first TTF surface: {}", SDL_GetError());
        throw std::runtime_error(error);
    }

    // Create target surface with increased size for padding.
    int padding = radius * 2;
    SDL_Surface *target_surf = SDL_CreateRGBSurface(0,text_surf->w + padding,text_surf->h + padding,32,0,0,0,0);
    if (!text_surf) {
        auto error = fmt::format("Error creating target text surface: {}", SDL_GetError());
        throw std::runtime_error(error);
    }

    this->rect.w = text_surf->w;
    this->rect.h = text_surf->h;

    // Using the outer_color text to stamp in a circle.

    // Polar Coordinates trigonometry Algorithm
    // for (int index = 0; index < (int)(M_PI * radius); index++) {
    //     double x = (cos(index / (radius / 2)) * radius) + radius;
    //     double y = (sin(index / (radius / 2)) * radius) + radius;
    //     this->rect.x = x; this->rect.y = y;
    //     SDL_BlitSurface(text_surf, NULL, target_surf, &this->rect);
    // }

    // Bresenham's Circle Drawing Algorithm
    // https://www.geeksforgeeks.org/bresenhams-circle-drawing-algorithm/
    int x = 0;
    int y = radius;
    int d = radius;
    while (y >= x) {
        this->rect.x = radius + x; this->rect.y = radius + y;
        SDL_BlitSurface(text_surf, NULL, target_surf, &this->rect);
        this->rect.x = radius + x; this->rect.y = radius - y;
        SDL_BlitSurface(text_surf, NULL, target_surf, &this->rect);
        this->rect.x = radius - x; this->rect.y = radius + y;
        SDL_BlitSurface(text_surf, NULL, target_surf, &this->rect);
        this->rect.x = radius - x; this->rect.y = radius - y;
        SDL_BlitSurface(text_surf, NULL, target_surf, &this->rect);
        this->rect.x = radius + y; this->rect.y = radius + x;
        SDL_BlitSurface(text_surf, NULL, target_surf, &this->rect);
        this->rect.x = radius + y; this->rect.y = radius - x;
        SDL_BlitSurface(text_surf, NULL, target_surf, &this->rect);
        this->rect.x = radius - y; this->rect.y = radius + x;
        SDL_BlitSurface(text_surf, NULL, target_surf, &this->rect);
        this->rect.x = radius - y; this->rect.y = radius - x;
        SDL_BlitSurface(text_surf, NULL, target_surf, &this->rect);
        x++;
        if (d > 0) {
            y--;
            d = d + 4 * (x - y) + 10;
        } else {
            d = d + 4 * x + 6;
        }
    }

    // Free text_surf before using it to create center text.
    SDL_FreeSurface(text_surf);
    text_surf = NULL;

    // Create Center Text center_color.
    text_surf = TTF_RenderText_Blended(font, text, center_color);
    if (!text_surf) {
        auto error = fmt::format("Error creating second TTF surface: {}", SDL_GetError());
        throw std::runtime_error(error);
    }

    // Using the center_color stamp the text in the center.
    this->rect.x = radius; this->rect.y = radius;
    SDL_BlitSurface(text_surf, NULL, target_surf, &this->rect);

    // Create a gpu accelerated texture image from the target surface.
    this->image = SDL_CreateTextureFromSurface(this->renderer, target_surf);
    if (!this->image) {
        auto error = fmt::format("Error creating a text texture: {}", SDL_GetError());
        throw std::runtime_error(error);
    }

    // Get a rect from the final texture image.
    if (SDL_QueryTexture(this->image, NULL, NULL, &this->rect.w, &this->rect.h)) {
        auto error = fmt::format("Error while querying texture: {}", SDL_GetError());
        throw std::runtime_error(error);
    }

    this->rect.x = 0;
    this->rect.y = 0;

    // free memory allocated for surfaces and font.
    SDL_FreeSurface(text_surf);
    SDL_FreeSurface(target_surf);
    TTF_CloseFont(font);
}

void BubbleText::update() {
    this->rect.x += this->vel_x;
    if (this->vel_x > 0) {
        if (this->rect.x + this->rect.w > Game::width) {
            this->vel_x = -2;
        }
    } else {
        if (this->rect.x < 0) {
            this->vel_x = 2;
        }  
    }
    this->rect.y += this->vel_y;
    if (this->vel_y > 0) {
        if (this->rect.y + this->rect.h > Game::height) {
            this->vel_y = -2;
        }
    } else {
        if (this->rect.y < 0) {
            this->vel_y = 2;
        }  
    }
}

BubbleText::~BubbleText() {
    SDL_DestroyTexture(this->image);
}

void BubbleText::draw() {
    if (SDL_RenderCopy(this->renderer, this->image, nullptr, &this->rect)) {
        auto error = fmt::format("Error while rendering texture: {}", SDL_GetError());
        throw std::runtime_error(error);
    }
}

File: src/bubble_text.rs
use sdl2::pixels::{Color, PixelFormatEnum};
use sdl2::rect::Rect;
use sdl2::render::{Texture, TextureCreator, WindowCanvas};
use sdl2::surface::Surface;
use sdl2::ttf::Sdl2TtfContext;
use sdl2::video::WindowContext;
// use std::f64::consts::PI;

use crate::{SCREEN_WIDTH, SCREEN_HEIGHT};

pub struct BubbleText<'a> {
    texture: Texture<'a>,
    rect: Rect,
    vel_x: i32,
    vel_y: i32
}

impl<'a> BubbleText<'a> {
    pub fn new(
        texture_creator: &'a TextureCreator<WindowContext>,
        ttf_context: &'a Sdl2TtfContext,
        text: String,
        size: u16,
        radius: f64,
        outer_color: Color,
        center_color: Color
    ) -> Result<BubbleText<'a>, String> {

        let font = ttf_context.load_font("fonts/freesansbold.ttf", size)?;

        let outer_surf = font.render(&text).solid(outer_color).map_err(|e| e.to_string())?;
        let center_surf = font.render(&text).blended(center_color).map_err(|e| e.to_string())?;
        let mut dest_rect = Rect::new(
            radius as i32,
            radius as i32,
            outer_surf.width(),
            outer_surf.height()
        );

        let mut target_surf = Surface::new(
            outer_surf.width() + (radius as u32) * 2,
            outer_surf.height() + (radius as u32) * 2,
            PixelFormatEnum::ARGB8888
        ).map_err(|e| e.to_string())?;

        // Using the outer_color text to stamp in a circle.

        // Polar Coordinates trigonometry Algorithm
        // for index in 0..(PI * radius) as i32 {
        //     let x = (f64::cos(index as f64 / (radius / 2.0)) * radius) + radius;
        //     let y = (f64::sin(index as f64 / (radius / 2.0)) * radius) + radius;
        //     dest_rect.x = x as i32;
        //     dest_rect.y = y as i32;
        //     outer_surf.blit(None, &mut target_surf, Some(dest_rect))?;
        // }

        // Bresenham's Circle Drawing Algorithm
        // https://www.geeksforgeeks.org/bresenhams-circle-drawing-algorithm/
        let mut x = 0;
        let mut y = radius as i32;
        let mut d = 3 - 2 * (radius as i32);
        blit(&outer_surf, &mut target_surf, &mut dest_rect, radius as i32, x, y)?;
        while y >= x {
            x += 1;
            if d > 0 {
                y -= 1;
                d = d + 4 * (x - y) + 10;
            } else {
                d = d + 4 * x + 6;
            }
            blit(&outer_surf, &mut target_surf, &mut dest_rect, radius as i32, x, y)?;
        }

        dest_rect.x = radius as i32;
        dest_rect.y = radius as i32;
        center_surf.blit(None, &mut target_surf, Some(dest_rect))?;

        let texture = texture_creator.create_texture_from_surface(&target_surf).map_err(|e| e.to_string())?;
        let rect = Rect::new(0, 0, target_surf.width(), target_surf.height());
        let vel_x = 2;
        let vel_y = 2;
        Ok(BubbleText { texture, rect, vel_x, vel_y})
    }

    pub fn draw(&self, canvas: &mut WindowCanvas) -> Result<(), String> {
        canvas.copy(&self.texture, None, self.rect)?;
        Ok(())
    }

    pub fn update(&mut self) {
        self.rect.x += self.vel_x;
        if self.vel_x > 0 {
            if self.rect.x + self.rect.w > SCREEN_WIDTH as i32 {
                self.vel_x = -2;
            }
        } else {
            if self.rect.x < 0 {
                self.vel_x = 2;
            }  
        }

        self.rect.y += self.vel_y;
        if self.vel_y > 0 {
            if self.rect.y + self.rect.h > SCREEN_HEIGHT as i32 {
                self.vel_y = -2;
            }
        } else {
            if self.rect.y < 0 {
                self.vel_y = 2;
            }  
        }
    }
}

fn blit(
    outer_surf: &Surface,
    target_surf: &mut Surface,
    dest_rect: &mut Rect,
    radius: i32,
    x: i32,
    y: i32
) -> Result<(), String> {
    dest_rect.x = radius + x; dest_rect.y = radius + y;
    outer_surf.blit(None, target_surf, Some(*dest_rect))?;
    dest_rect.x = radius - x; dest_rect.y = radius + y;
    outer_surf.blit(None, target_surf, Some(*dest_rect))?;
    dest_rect.x = radius + x; dest_rect.y = radius - y;
    outer_surf.blit(None, target_surf, Some(*dest_rect))?;
    dest_rect.x = radius - x; dest_rect.y = radius - y;
    outer_surf.blit(None, target_surf, Some(*dest_rect))?;
    dest_rect.x = radius + y; dest_rect.y = radius + x;
    outer_surf.blit(None, target_surf, Some(*dest_rect))?;
    dest_rect.x = radius - y; dest_rect.y = radius + x;
    outer_surf.blit(None, target_surf, Some(*dest_rect))?;
    dest_rect.x = radius + y; dest_rect.y = radius - x;
    outer_surf.blit(None, target_surf, Some(*dest_rect))?;
    dest_rect.x = radius - y; dest_rect.y = radius - x;
    outer_surf.blit(None, target_surf, Some(*dest_rect))?;
    Ok(())
}